### ISSUE TITLE
"dotnet run" went from 4 seconds to 45 seconds to lauch project

### ISSUE THREAD

**Issue by khowling** [dotnet/cli#7551]
Looks like dotnet is scanning the complete directory rather than specifying a 'launch' .cs file and following dependencies used from that file.  Equally, it would be good to be able to specify different launch commands under one git repo to manage different process types, for example: 

```
> dotnet run worker_imageprocessor.cs
> dotnet run kestel_server.cs
```

## Steps to reproduce

Working on a small app, c#  backend, react frontend.   Used 'create-react-app & npm install' to create a folder 'frontend' under project directory (common directory structure in nodejs projects)

## Expected  behavior

Expected this would have no impact on dotnet cli as the frontend folder has no files related to the backend build process, however.

## Actual behaviour

It caused every dotnet run to take 10x longer.  

## Environment data

.NET Command Line Tools (2.0.0)

Product Information:
 Version:            2.0.0
 Commit SHA-1 hash:  cdcd1928c9

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.0\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d

**Comment by benaadams** [comment]
Try adding excludes to `.csproj` for node_modules https://github.com/dotnet/cli/issues/7525#issuecomment-324998276

e.g.

```
<PropertyGroup>
  <DefaultItemExcludes>node_modules/**;$(DefaultItemExcludes)</DefaultItemExcludes>
</PropertyGroup>
```

**Comment by sharwell** [comment]
:bulb: @benaadams Typically these would be ordered so the new items are added to the end. In reality it doesn't make a functional difference, but it a bit easier to read (since this one is different, I did a double-take to see if it was still correct).

```xml
<DefaultItemExcludes>$(DefaultItemExcludes);node_modules/**</DefaultItemExcludes>
```

**Comment by dasMulli** [comment]
@sharwell No, the order does matter in a few cases, especially since the default exclude patterns contains `**/node_modules/**` which triggers an msbuild perf bug when it is **before** `foo/node_modules/**` (the directory is excluded, but msbuild walks through the whole directory)

**Comment by peterhuene** [comment]
I have a fix (#8907) that prevents "double globbing": once for the build and again for the run property evaluation.  For directories containing many files, such as (potentially) `node_modules`, this second glob contributed to poor performance, especially when the first glob is skipped entirely (i.e. `dotnet run --no-build`).

`DefaultItemExcludes` should still be used used to prevent the build from globbing unnecessarily, but with that fix, run will not glob for any files, regardless of `DefaultItemExcludes` being set.

**Comment by peterhuene** [comment]
The fix is checked in.  This should improve performance of `dotnet run --no-build` considerably and reduce the overall time of just `dotnet run` by a considerable amount when run in directories containing a large set of files, even without `DefaultItemExcludes` set.

I still recommend setting that property to exclude unnecessary directories from being globbed for the build, though.
